asllib/ASLTypingReference.mld

```diff
@@ -16,107 +16,753 @@ welcome; those can be sent to Arm’s Architecture Formal Team Lead Jade
+        - S is an anonymous type;
+        - T is a named type;
+        - ty is T.
+
+     - All of the following applies:
+        - S is an anonymous type;
+        - T is an anonymous type;
+	- ty is the well-constrained integer type whose domain is the union of the
+	  domains of S and T.      
+
+  - All of the following applies:
+    - Either S or T have the structure of an unconstrained integer type;
+    - One of the following applies:
+
+      - All of the following applies:
+        - S is a named type;
+        - S has the structure of an unconstrained integer type;
+        - T is an anonymous type;
+        - ty is S.
+
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is a named type;
+        - T has the structure of an unconstrained integer type;
+        - ty is T.
+
+
+      - All of the following applies:
+        - S is an anonymous type;
+        - T is an anonymous type;
+	- ty is the unconstrained integer type. 
+
+  - All of the following applies:
+    - Either S or T have the structure of an under-constrained integer type;
+    - ty is the under-constrained integer type. 
+
+  - ty is undefined.
+
+  {3 Comments}
+  This encompasses LRM Section 7.6.1. R_YZHM.
+
+{2 TypingRule.CheckUnop}
+
+  {3 Goal}
+  Checking compatibility of an unary operator with the type of its argument.
+
+  {3 Definition}
+  [t] is the result of checking compatibility of a unary operator [op] with
+  type [t1] and one of the following applies:
+  - All of the folowing applies:
+    - [op] is [BNOT];
+    - [t1] type-satisfies [boolean];
+    - [t] is [boolean];
+
+  - All of the following applies:
+    - [op] is [NEG];
+    - One of the following applies:
+      - [t1] type-satisfies [integer]; 
+      - [t1] type-satisfies [real];
+     - One of the following applies:
+       - All of the following applies:
+         - [t1] has the structure of an unconstrained integer;
+         - [t] is an unconstrained integer;
+       - All of the following applies:
+         - [t1] has the structure of a constrained integer;
+         - [t] is a constrained integer whose constraint is ;
+  
+  - All of the following applies:
+    - [op] is [NOT];
+    - [t1] has the structure of a bitvector;
+    - [t] is [t1].
+
+  {3 Comments}
+
+  {3 Examples}
+
+ {2 TypingRule.CheckBinop}
+
+  {3 Goal}
+  Checking compatibility of a binary operator with the types of its arguments.
+
+  {3 Definition}
+  [t] is the result of checking compatibility of a binary operator [op] with
+  types [t1] and [t2] and one of the following applies:
+  - All of the following applies:
+    - [op] is [AND], [OR], [EQ] or [IMPL]; 
+    - [t1] type-satisfies [boolean];
+    - [t2] type-satisfies [boolean];
+    - [t] is [boolean].
+
+  - All of the following applies:
+    - [op] is [AND], [OR], or [EOR];
+    - [t1] has the structure of a bitvector;
+    - [t2] has the structure of a bitvector;
+    - [t1] and [t2] have the same bitvector width [w];
```

Comment:
This (and the following) would also benefit from a definition of equality for bitvector width.